### PR TITLE
Parent row ID that we discussed in #553

### PR DESCRIPTION
**Pull request by asmirn1**
Hi Tomas,

These are changes to store ID of a parent row. A few comments about my changes:

1. As I understood our discussion in #553, we have decided to use the field ID of PeakListRow as an identifier of a parent feature (chromatogram). In this case, I believe that it makes sense to store the parent ID as a property of PeakListRow instead of Feature. Let me know if you want to change it.

2. I've noticed that there are several versions of ProjectLoad... methods. I've changed the latest version (2_5). I'm not sure if I should change the older versions or make a new version for my changes?

3. I'll push changes to spectral deconvolution that are using this new feature in a separate PL.

Alex.

### CODE REVIEW

**Comment by tomas-pluskal** [comment]
Hi Alex,

The parent chromatogram ID should be a property of Feature, otherwise you will create a mess after alignment.

Updating the latest ProjectLoad.. module is fine.

Since you added this parent ID, it would be nice if the user has a way to see it somehow. E.g., you can add it to the peak list table visualizer as a new DataFileColumnType.

Please also make sure the parent ID is correctly retained after deisotoping, alignment, gap filling, etc.

File: src/main/java/net/sf/mzmine/datamodel/impl/SimplePeakListRow.java [hunk 1]
```diff
@@ -52,6 +54,7 @@
   private PeakInformation information;
   private int myID;
   private double maxDataPointIntensity = 0;
+  private Integer parentRowID;
```
Comment (tomas-pluskal) [inline src/main/java/net/sf/mzmine/datamodel/impl/SimplePeakListRow.java]: Please use the ```@Nullable``` annotation

File: src/main/java/net/sf/mzmine/datamodel/impl/SimplePeakListRow.java [hunk 2]
```diff
@@ -413,5 +416,15 @@ public void update() {
   // End Gauthier edit
 
 
+  @Override
+  public void setParentRowID(Integer id) {
```
Comment (tomas-pluskal) [inline src/main/java/net/sf/mzmine/datamodel/impl/SimplePeakListRow.java]: Please use the ```@Nullable``` annotation

**Comment by asmirn1** [comment]
Hi Tomas,

Regarding DataFileColumnType. I see how you create columns in class `PeakListTable`, but I couldn't find the place where table cells are filled out with actual values from a peaklist. Could you tell me what class I need to change to fill my new column with values?

**Comment by tomas-pluskal** [comment]
Hi Alex,
check the ```PeakListTableModel.getValueAt()``` method.

**Comment by asmirn1** [comment]
Hi Tomas,

> Please also make sure the parent ID is correctly retained after deisotoping, alignment, gap filling, etc

If we continue this list of modules, it may include almost all peaklist methods in MZmine. I don't think it's a good idea to change all of them. So before I'll start making changes, could we come up with a short list of modules that need to be changed?

My original idea was to make the minimum number of changes so that:
1. The ADAP workflow (i.e. **Chromatogram builder** -> **Chromatogram deconvolution** -> **Spectral deconvolution** -> **Alignment**) would work as it is supposed to work, and
2. My changes won't break any other modules, algorithms, and workflows.

Therefore, I was planning to change only **Chromatogram deconvolution** module to save the chromatogram ID and **Spectral deconvolution** to read that ID. 

I will also make sure that **Spectral deconvolution** works even when the chromatogram ID is missing (i.e. `null`). So if users don't follow the ADAP workflow, they'll still be able to do spectral deconvolution, but they will not have that feature of merging adjacent peaks from the same chromatogram.

This way I will avoid changing everybody's code and potentially breaking something.

If this doesn't work for you and you would like to have changes in other modules, let's make a list of such modules.

**Comment by tomas-pluskal** [comment]
Hi Alex,

I think most of the downstream modules (like deisotoping) create instances of SimpleFeature with a copy constructor:

https://github.com/mzmine/mzmine2/blob/master/src/main/java/net/sf/mzmine/modules/peaklistmethods/isotopes/deisotoper/IsotopeGrouperTask.java#L209

So all you need to do is to modify the copy constructor in SimpleFeature and make sure the parent row ID gets propagated to the new instance.

**Comment by asmirn1** [comment]
Okay. I'll try that.

**Comment by asmirn1** [comment]
Hi Tomas,

I've made the following changes:

- Remove ParentRowID from PeakListRow.
- Add ParentChromatographyRowID to SimpleFeature and ResolvedPeak
- Replace Feature[] with ResolvedPeak[] in the deconvolution methods to be able to use method setParentChromatographyRowID() of ResolvedPeak

Could you take a look at my changes.

Alex.

**Comment by tomas-pluskal** [comment]
Hi Alex, 
Looks good, thanks!

**Comment by asmirn1** [comment]
Thank you Tomas! 

I'll make PR with corresponding changes to the spectral deconvolution soon.

**Pull request merged**